### PR TITLE
Refactor torchao docs 

### PR DESCRIPTION
**Pull request by MekkCyber**
# What does this PR do?
Refactoring torchao docs, with clear doc layout, better and exhaustive examples, benchmarks and more !

### CODE REVIEW

**Comment by SunMarc** [comment]
cc @DerekLiu35

File: docs/source/en/quantization/torchao.md [hunk 1]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
+| Feature | Description |
+|--------|-------------|
+| **Post-Training Quantization (PTQ)** | Quantize weights and activations  with just one line, see below for examples |
+| **Quantization Aware Training (QAT)** | Train quantized models with minimal accuracy loss, for more details see [QAT README](https://github.com/pytorch/ao/blob/main/torchao/quantization/qat/README.md) |
+| **Float8 Training** | High-throughput training with float8 formats, for more details see [torchtitan](https://github.com/pytorch/torchtitan/blob/main/docs/float8.md) |
+| **Sparsity Support** | Semi-structured (2:4) sparsity for faster inference, details in the [blog](https://pytorch.org/blog/accelerating-neural-network-training/) |
+| **Optimizer Quantization** | Reduce optimizer state memory with 4/8-bit variants of Adam. |
+| **KV Cache Quantization** | Enables long context inference with lower memory, some details [here](https://github.com/pytorch/ao/blob/main/torchao/_models/llama/README.md) |
+| **Custom Kernels Support** | use your own ops compatible with `torch.compile`. |
+| **FSDP2** | Composable with FSDP2 for training|
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:27]: Maybe we should put that at the bottom for additional features that people can test using torchao ? Since they are on the quantization docs, I feel like we should first show them how to quantize a model with torchao.

File: docs/source/en/quantization/torchao.md [hunk 2]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
 
+TorchAO supports several quantization techniques:
 
-Create a [`TorchAoConfig`] and specify the quantization type and `group_size` of the weights to quantize. Set the `cache_implementation` to `"static"` to automatically [torch.compile](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) the forward method.
+- **Int8 Weight Only**
+- **Int4 Weight Only**
+- **Int8 Dynamic Quantization**
+- **Float8 (E4M3 / E5M2)**
+- **KV Cache Quantization**
+- **Custom intx**: Support for int2, int3, etc., via pure PyTorch bitpacking (alpha feature)
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:46]: Maybe put that above the hardware compatibility section

File: docs/source/en/quantization/torchao.md [hunk 3]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
+## Supported Quantization techniques
 
+TorchAO supports several quantization techniques:
 
-Create a [`TorchAoConfig`] and specify the quantization type and `group_size` of the weights to quantize. Set the `cache_implementation` to `"static"` to automatically [torch.compile](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) the forward method.
+- **Int8 Weight Only**
+- **Int4 Weight Only**
+- **Int8 Dynamic Quantization**
+- **Float8 (E4M3 / E5M2)**
+- **KV Cache Quantization**
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:45]: not related to model quantization, maybe we can put that somewhere else

File: docs/source/en/quantization/torchao.md [hunk 4]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
-Starting with version 0.10.0, torchao provides enhanced flexibility through the `AOBaseConfig` API, allowing for more customized quantization configurations.
-And full access to the techniques offered in the torchao library.
+## Hardware Compatibility
 
-You can manually choose the quantization types and settings or automatically select the quantization types.
+| Component | Compatibility |
+|----------|----------------|
+| **CUDA Versions** | ✅ cu118, cu124, cu126, cu128 |
+| **CPU** | ✅ Quantization on cpu is supported (different layouts for Int4, see below) |
+| **CPU / GPU Offloading** | ✅ Optimizers and gradients can be offloaded |
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:35]: I don't think many user will understand this so let's remove that and maybe dedicate a seperate section about that if it make sense
```suggestion
```

File: docs/source/en/quantization/torchao.md [hunk 5]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
 
-```py
-import torch
-from transformers import TorchAoConfig, AutoModelForCausalLM, AutoTokenizer
-from torchao.quantization import Int4WeightOnlyConfig
-
-# Using AOBaseConfig instance (torchao >= 0.10.0)
-quant_config = Int4WeightOnlyConfig(group_size=128)
-quantization_config = TorchAoConfig(quant_type=quant_config)
+# Use in transformers
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:50]: no need to say that this is for transformers
```suggestion
```

File: docs/source/en/quantization/torchao.md [hunk 6]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
-output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implementation="static")
-print(tokenizer.decode(output[0], skip_special_tokens=True))
+Stable release from Pypi which will default to CUDA 12.4
+```bash
+# Updating 🤗 Transformers to the latest version, as the example script below uses the new auto compilation
+pip install --upgrade torchao transformers
+```
+Stable Release from the PyTorch index
+```bash
+pip install torchao --extra-index-url https://download.pytorch.org/whl/cu124 # options are cpu/cu118/cu124/cu126
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:63]: so doing pip install torchao will install the cu124 version ?

File: docs/source/en/quantization/torchao.md [hunk 7]
```diff
@@ -11,56 +11,56 @@ rendered properly in your Markdown viewer.
...
 
-tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3-8B")
-input_text = "What are we having for dinner?"
-input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+Install torchao with the following command.
 
-# auto-compile the quantized model with `cache_implementation="static"` to get speed up
-output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implementation="static")
-print(tokenizer.decode(output[0], skip_special_tokens=True))
+Stable release from Pypi which will default to CUDA 12.4
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:56]: maybe put that just above the bash command for torchao

File: docs/source/en/quantization/torchao.md (hunk 3 above)
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:45]: maybe we can just talk about the methods that we support, this will be easier. Otherwise, we will have a constantly update the docs.

**Review by SunMarc — COMMENTED** [review]
Thanks for reorganizing the docs ! Left a couple of comments !

File: docs/source/en/quantization/torchao.md (hunk 6 above)
Comment (MekkCyber) [reply docs/source/en/quantization/torchao.md:63]: Yup, written in their docs : https://github.com/pytorch/ao?tab=readme-ov-file#installation

File: docs/source/en/quantization/torchao.md [hunk 8]
```diff
@@ -106,76 +89,256 @@ For a complete list of available configurations, see our [quantization API docum
...
+input_text = "What are we having for dinner?"
+input_ids = tokenizer(input_text, return_tensors="pt")
+
+# auto-compile the quantized model with `cache_implementation="static"` to get speed up
+output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implementation="static")
+print(tokenizer.decode(output[0], skip_special_tokens=True))
+```
+</hfoption>
+
 <hfoption id="automatic">
```
Comment (DerekLiu35) [inline docs/source/en/quantization/torchao.md:322]: 🔥 LGTM! just one suggestion
```suggestion
<hfoption id="autoquant">
```
Maybe change automatic to autoquant

Comment (MekkCyber) [reply docs/source/en/quantization/torchao.md]: Thanks ! Done

File: docs/source/en/quantization/torchao.md [hunk 9]
```diff
@@ -199,30 +451,36 @@ def benchmark_fn(func: Callable, *args, **kwargs) -> float:
...
+## Other Features proposed by torchao
 
-To avoid arbitrary user code execution, torchao sets `weights_only=True` in [torch.load](https://pytorch.org/docs/stable/generated/torch.load.html) to ensure only tensors are loaded. Any known user functions can be whitelisted with [add_safe_globals](https://pytorch.org/docs/stable/notes/serialization.html#torch.serialization.add_safe_globals).
+| Feature | Description |
+|--------|-------------|
+| **Quantization Aware Training (QAT)** | Train quantized models with minimal accuracy loss, for more details see [QAT README](https://github.com/pytorch/ao/blob/main/torchao/quantization/qat/README.md) |
+| **Float8 Training** | High-throughput training with float8 formats, for more details see [torchtitan](https://github.com/pytorch/torchtitan/blob/main/docs/float8.md) and [accelerate](https://huggingface.co/docs/accelerate/usage_guides/low_precision_training#configuring-torchao) |
+| **Sparsity Support** | Semi-structured (2:4) sparsity for faster inference, details in the [blog](https://pytorch.org/blog/accelerating-neural-network-training/) |
+| **Optimizer Quantization** | Reduce optimizer state memory with 4/8-bit variants of Adam. |
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:471]: we should add docs in optimizer about 4-8bit adam from torchao

File: docs/source/en/quantization/torchao.md [hunk 10]
```diff
@@ -11,56 +11,39 @@ rendered properly in your Markdown viewer.
...
-You can manually choose the quantization types and settings or automatically select the quantization types.
+- **Int8 Weight Only**
+- **Int4 Weight Only**
+- **Int8 Dynamic Quantization**
+- **Float8 (E4M3 / E5M2)**
+- **AutoQuant**
 
-<hfoptions id="torchao">
-<hfoption id="manual">
+## Hardware Compatibility
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:27]: remove

**Review by SunMarc — APPROVED** [review]
Thanks ! would be nice to have a button called try on colab !

**Review by stevhliu — COMMENTED** [review]
Very nice and thorough refactor, thanks so much! 🔥 

I think it'd also be nice to create a separate section for the Migration and deprecation sections after `## Loading quantized models`. You can add a note in the installation section to refer to the Migration section if you're using an older version.

**Review by SunMarc — APPROVED** [review]
Thanks !

**Comment by jerryzh168** [comment]
Thanks a lot @MekkCyber! I was also planning to add more examples recently, your PR makes it much easier, the formatting is really nice
